README.md
# Moomoo.io AI Bot Sender

This script is a very configurable botting tool for the MMO game [Moomoo.io](http://moomoo.io).

### WARNING: Please do not download any code that claims that it will "fix" or "add features" to this bot code. People have been including trojans and viruses in those unofficial programs. The ONLY official Moomoo bot code authors are `Mega-Mewthree` and `Nebula-Developers`. DO NOT download any other code related to Moomoo bots.

We are not responsible for any damages caused by this project, or derivatives of this project. Use at your own risk.

## Resources

* [Discord](https://discord.gg/VgKFcVf)
* [Subreddit](https://reddit.com/r/Nebula_Devs)

For contributors: [To-Do List](https://github.com/Nebula-Developers/Moomoo-AI-Bot-Sender/projects/1)

## Installation

**NEW VIDEO INSTALLATION GUIDE:** https://www.youtube.com/watch?v=6QfVIt5V4oI

First, you must download and install Node.js: https://nodejs.org/en/

For all the Windows users who have no clue how to open a Command Prompt terminal/console, open `openCommandPrompt.bat`.

### Install Compilers to Make Follow Mouse Command Work (Optional)

If you have already installed the bot code, run `npm uninstall robotjs` in your console/terminal inside of the bot code's directory.

#### Windows

Download and install Visual C++ Build Tools 2015 [here](http://landinghub.visualstudio.com/visual-cpp-build-tools).

Download and install Python 2.7.x [here](https://www.python.org/downloads/release/python-2714/). Make sure you select the option to add it to PATH.

Run the following commands in your console (Command Prompt/Powershell):

    npm config set python python2.7
    npm config set msvs_version 2015

#### Mac OS X

Download and install Python 2.7.x [here](https://www.python.org/downloads/release/python-2714/) if it isn't already installed.

Download and install Xcode [here](https://developer.apple.com/xcode/download/).

#### Unix

You need Python 2.7.x and `make`.

Download and install GCC [here](https://gcc.gnu.org/).

#### All Systems

Run the following command in your console/terminal inside of the bot code's directory:

    npm install robotjs

### Install the Bot Code (Required)

After Node.js is installed, run the below command in your console inside this folder. It is required to install important code the script depends on to run, and you can do that easily with the following command:

    npm install

Or you can run `npminstall.bat` if you are on Windows and you have no clue how to get to a console.

Install a userscript manager:

Chrome: [Tampermonkey](https://chrome.google.com/webstore/detail/tampermonkey/dhdgffkkebhmkfjojejmpbldmpobfkfo?hl=en)

Firefox: [Tampermonkey](https://addons.mozilla.org/firefox/addon/tampermonkey/)

Safari: [Tampermonkey](http://tampermonkey.net/?browser=safari)

Microsoft Edge: [Tampermonkey](https://www.microsoft.com/store/p/tampermonkey/9nblggh5162s)

Opera: [Tampermonkey](https://addons.opera.com/extensions/details/tampermonkey-beta/)


Put the code in showid.user.js into Tampermonkey so that your ID is visible.


Here is the start script:

    node index.js [--probeTribe tribeName] [--probeName playerName] [--num numberOfBots] [--link moomooPartyLink] [--tribe tribeName] [--name botName] [--randNames true] [--chat chatMessage] [--ai true] [--autoHeal true] [--randSkins true]

See the next section for information on everything after `node index.js`.

## Configuration

There are two main modes to this bot, probing and botting.

#### NEW: Open start.bat to get an easier to use UI! For people who can't use the command line very well! (Windows only)

### Botting

Type "setowner \<your ID>" into the node.js console to make the bot obey your commands.

`--num` sets the number of bots. Required for bots to function, and servers have their own limit of around 20 bots and users. Default: `0`

`--link` sets the Moomoo party link. Required for bots to function. Default: none

`--tribe` sets the Moomoo tribe. Default: none

`--name` sets the bots' names. Default: `unknown`

`--randNames` selects random names for the bots. Default: `false`

`--chat` sets the bots' chat message. Default: none

`--ai` enables using commands. If AI is not enabled, the bots spawn and do nothing. Default: `false`

`--autoHeal` enables auto healing. Default: `true`

`--randSkins` randomizes the skin color of bots. Default: `false`

`--hat` changes the hat via ID, name, or keyword, that will be bought/equipped when enough gold is reached. You can find each hat's ID online, and soon in the Tampermonkey script. Default: none

`--autoAttack` enables auto attack (as if they pressed E). Default: `true`

### Probing

`--probeTribe` searches all Moomoo.io servers for any tribes with the given name and returns the IPs so you can use them with --link.

`--probeName` does almost the same thing as --probeTribe, but searches the leaderboards for a player with the given name.

`--probeRegex` makes the arguments provided to the 2 previous probe flags be interpreted as a regular expression. Do not use flags in your regular expression. Note that the starting and ending slashes (`/`) are not required. It is suggested that you surround your search expression in quotation marks so that the terminal doesn't interpret special characters like `>` as command line operators. Default: `false`

You can probe (search) for a name on the leaderboard or a tribe on all the public Moomoo.io servers. This is useful if you want to use the other part of this script to put bots on someone's livestream.

Here is the start script for probing using a tribe name:

    node index.js --probeTribe Nebula

This will return a list of server IPs which have the tribe `Nebula` in them.

Here is the start script for probing using a name on the leaderboard:

    node index.js --probeName Lucario

Similar to the other command, this will return a list of server IPs which have the player `Lucario` on the leaderboard.

Here is an example of regular expression probing:

    node index.js --probeName "Luc" --probeRegex true

This will return all servers where someone whose name has the string `Luc` is on the leaderboard. For example, it will match `Luc`, `Luca`, `Lucario`, and `abcLucdef`. (But not `luc`.)

You can put both the outputs into `--link` for the botting start script.

## Console Commands

Type `setowner <your ID>` into the Node.js console to make the bot obey your ingame commands.

## Ingame Commands

Press R or click on the minimap to make a ping on the minimap, which makes the bots come to that pinged location.

!id \<player name>: Makes the bots send the ID of a player in chat.

!fme: Makes the bots follow you.

!fid \<player ID>: Makes the bots follow the player with that ID.

!atkid \<player ID>: Makes the bots attack the player with that ID. [BUGGY]

!s: Makes the bots stay.

!r: Releases the bots from control. Different from !s in that pressing R to ping will move them.

!fm: Makes the bots follow your mouse. May be buggy if the robot.js lib didn't compile correctly.

!hat \<hat ID>: Switches a bot's hat using the ID, name, or keyword. If necessary, the requested hat will be bot.

!atk: Toggles auto attack, as if the bots pressed E.


changelog.txt
June 28, 2018 (2.1.0):
  + Bots are back
    * There will still be a warning when using 20 or more bots though
    * No known way to bypass IP limit other than proxies (might come eventually)
  + You can re-enable your script, it has some useful features

June 27, 2018 (2.0.0):
  + Sidney added an IP limit to Moomoo.io
    * RIP bots 2017-2018
    * This script can now only probe servers.
  + Updated the server list to fetch from server.
  + Probing still works, so raid livestreams (manually) all you want.
  + Feel free to disable the Tampermonkey script now.
  + It's been a nice run everyone!
  + Thank you for using this script!
  + Join our Discord at https://nebula.mooo.info

April 22, 2018 (1.9.0):
  + JOIN NEBULA AT https://join-nebula.surge.sh (Discord)
  + Added regex probing
    * See the documentation for details
  + Fixed some errors when sending large numbers of bots

April 14, 2018 (1.8.1):
  + Added dev servers

March 29, 2018 (1.8.0):
  + Updated Moomoo Bot Utilities to work with the new encoding scheme
  + UPDATE YOUR TAMPERMONKEY SCRIPT!!!

February 23, 2018 (1.7.0):
  + Updated code to use ArrayBuffers
  + The first time you run this, "npm install" will be automatically run
  + "npm install" will be run automatically after any future updates

February 19, 2018 (1.6.1):
  + Fixed a typo that crashes the program whenever a tribe is deleted

February 12, 2018 (1.6.0):
  + NEW NEBULA SERVER!!!: https://discord.gg/Dvqy2A8
  + Follow mouse now automatically disables after respawning to allow pinging bots
  + Fixed some bugs with !hat
  + !sp to spawn spikes
  + !w to spawn walls
  + !atkid now spreads bots out if the last player update is >30 seconds ago
  + Version 2 coming soon!

February 9, 2018 (1.5.0):
  + Added --autoAttack for easier farming
  + Added !atk to toggle auto attack
  + Since I pushed 4 updates in an hour, here's a summary of the new changes:
    * !atkid is now fixed
    * !hat <hat name> will make your bots buy and equip that hat if it has enough gold
    * --hat <hat name> automatically buys and equips it
    * Tampermonkey script was updated to decrease conflict with other scripts
      - Update it if you use auto-heal or any other scripts

February 9, 2018 (1.4.2):
  + UPDATE YOUR TAMPERMONKEY SCRIPT!!!
  + This updated Tampermonkey script should have less conflicts with other scripts

February 9, 2018 (1.4.1):
  + Fixed an error when no hat is used

February 9, 2018 (1.4.0):
  + Bots can now buy and equip a hat (when a bot reaches enough gold to do so) using --hat <hat ID or hat name>
  + A new command, !hat, to make a bot attempt to buy/equip a hat
  + Fixed !atkid

January 31, 2018 (1.3.1):
  + Fixed a critical bug that crashed the code (I think)

January 31, 2018 (1.3.0):
  + Random skin colors with --randSkins
  + BREAKING CHANGE: --randnames renamed to --randNames

January 30, 2018 (1.2.2):
  + Fixed autoHeal flag in start.bat

January 28, 2018 (1.2.1):
  + Made autoheal default to true instead of false

January 28, 2018 (1.2.0):
  + No more need to manually setowner
  + UPDATE YOUR TAMPERMONKEY SCRIPT!!!

January 27, 2018 (1.1.1):
  + Added batch scripts to make using bots on Windows easier

January 27, 2018 (1.1.0):
  + Added auto updater


showid.js
// ==UserScript==
// @name         Moomoo Bot Utilities
// @namespace    https://discord.gg/Uj3GWPy
// @version      1.2.0
// @description  Shows your internal ID and position.
// @author       Mega_Mewthree
// @match        *://moomoo.io/*
// @match        *://45.77.0.81/*
// @match        *://dev.moomoo.io/*
// @grant        none
// @run-at       document-start
// ==/UserScript==

(() => {
    var ws = null;
    var id = null;
    var pos = [];

    WebSocket = class extends WebSocket {
        constructor(...arg) {
            super(...arg);
            ws = this;
            this.addEventListener('message', function(e){
                handleMessage(e);
            });
        }
    };

    function handleMessage(e){
        var m = e.data;
        if (!m.startsWith(`42["2",`) && !m.startsWith(`42["3",`) && !m.startsWith(`42["5",`) && !m.startsWith(`42["6",`)) displayID();
        if (m.startsWith(`42["1",`)){
            id = /(42\[\"1\",)([0-9]+)\]/.exec(m)[2];
            const req = new XMLHttpRequest();
            req.open("POST", `http://localhost:15729/?ownerID=${id}`, true);
            req.send();
        }else if (m.startsWith(`42["3",`)){
            var packet = m.replace(`42["3",`, "");
            packet = packet.substr(0, packet.length - 1);
            var data = JSON.parse(packet);
            for (var i = 0, len = data.length / 13; i < len; i++){
                if (id == data[0 + i * 13]){
                    pos[0] = data[1 + i * 13];
                    pos[1] = data[2 + i * 13];
                }
            }
        }
    }

    function displayID(){
        var t = document.getElementById("ageText");
        var age = /AGE [0-9]+/.exec(t.innerHTML);
        t.innerHTML = `${age && age[0]} (${id}) [${pos.join(", ")}]`;
    }

    setInterval(() => {
        if (id){
            const req = new XMLHttpRequest();
            req.open("POST", `http://localhost:15729/?ownerID=${id}`, true);
            req.send();
        }
    }, 5000);
})();


index.js
/*
Moomoo-AI-Bot-Sender
Copyright (C) 2018 Mega Mewthree

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

(async () => {
	const io = require("socket.io-client");
	const request = require("request");
	const url = require("url");
	const fs = require("fs");
	const { spawn } = require("child_process");

	let parser = null;
	try {
		require.resolve("socket.io-msgpack-parser");
		parser = require("socket.io-msgpack-parser");
	} catch(e) {
		console.error("Socket.io Msgpack Parser was not installed. Exiting and installing...");
		spawn("npm", ["install"], {
			stdio: "ignore",
			shell: true,
			detached: true
		});
		process.exit();
	}

	if (!fs.existsSync(`${__dirname}/lastUpdated.txt`) || Date.now() - parseInt(fs.readFileSync(`${__dirname}/lastUpdated.txt`, "utf8")) > 43200000) {
		spawn("node", [`${__dirname}/autoupdate.js`], {
			stdio: "ignore",
			shell: true,
			detached: true
		});
	}

	let computer = null;

	const screen = computer && computer.getScreenSize();

	const args = parseFlags(process.argv.slice(2).join(" "), ["--num", "--link", "--tribe", "--name", "--randNames", "--randSkins", "--chat", "--ai", "--probeTribe", "--probeName", "--probeRegex", "--autoHeal", "--hat", "--autoAttack"]);

	let ownerID = null;
	let followID = null;

	let attackFollowedPlayer = false;
	let stay = false;

	const goto = { x: null, y: null };

	let mousePos = { x: 0, y: 0 };
	let followMouse = false;
	let getMouseInputInterval = null;
	if (computer) {
		getMouseInputInterval = setInterval(() => {
			mousePos = computer.getMousePos();
		}, 200);
	}

	function get(url) {
		return new Promise((resolve, reject) => {
			request(url, (err, res, body) => {
				if (err) {
					reject(err);
					return;
				}
				if (body) {
					resolve(body);
				}
			});
		});
	}

	function parseFlags(string, flags_array) {
		if (!Array.isArray(flags_array)) {
			return { error: "Array of flags not found." };
		}
		const return_object = {};
		const flag_locations = [[-1, "null", []]];
		const string_array = string.split(" ");
		for (let index = 0; index < string_array.length; index++) {
			if (flags_array.indexOf(string_array[index]) > -1) {
				flag_locations.push([index, string_array[index], []]);
			}else{
				flag_locations[flag_locations.length - 1][2].push(string_array[index]);
			}
		}
		for (let index = 0; index < flag_locations.length; index++) {
			return_object[flag_locations[index][1].replace(/^(-*)/g, "")] = {};
			return_object[flag_locations[index][1].replace(/^(-*)/g, "")].flagLocation = flag_locations[index][0];
			return_object[flag_locations[index][1].replace(/^(-*)/g, "")].value = flag_locations[index][2].join(" ");
		}
		return return_object;
	}

	function getIP(link) {
		link = link.match(/\d+:\d+:\d+/g);
		if (link.length > 0) {
			return link[link.length - 1];
		} else {
			return false;
		}
	}

	function getServerByID(id) {
		let len = allServers.length;
		while (len--) {
			if (allServers[len].id === id) return allServers[len];
		}
		return false;
	}

	function getServerByIp(ip) {
		let len = allServers.length;
		while (len--) {
			if (allServers[len].ip === ip) return allServers[len];
		}
		return false;
	}

	function processInput(line) {
		const a = line.split(" ");
		const command = a.shift();
		if (command == "setowner") {
			ownerID = a[0];
		}
	}

	function getHatID(name) {
		const possiblyInt = parseInt(name);
		if (!isNaN(possiblyInt)) {
			return possiblyInt;
		} else {
			if (!name || !name.toString) return null;
			let safeName = name.toString().toLowerCase();
			safeName = safeName.replace(/[$-/:-?{-~!"^_`\[\]]/g, ""); // remove symbols
			safeName = safeName.replace(/\s/g, ""); // remove whitespace
			safeName = data.hatAliases[safeName];
			if (!isNaN(safeName)) return safeName;
			return null;
		}
	}

	const data = require("./data.json");

	const names = [
		"Wally",
		"Tanika",
		"Lenna",
		"Reid",
		"Joshua",
		"Miguelina",
		"Enda",
		"Ona",
		"Natashia",
		"Matt",
		"Shenika",
		"Tommye",
		"Corrin",
		"Angelyn",
		"Owen",
		"Zachariah",
		"Renata",
		"Shiloh",
		"Joesph",
		"Teresia",
		"Barabara",
		"Bee",
		"Janae",
		"Christel",
		"Tequila",
		"Becki",
		"Jacki",
		"Eboni",
		"Madge",
		"Elizabeth",
		"Ingeborg",
		"Latoyia",
		"Aretha",
		"Cecile",
		"Verdell",
		"Valda",
		"Pandora",
		"Alvina",
		"Tiara",
		"Kristopher",
		"Mikel",
		"Annita",
		"Concetta",
		"Reita",
		"Clarine",
		"Warner",
		"Stephani",
		"Herlinda",
		"Jeraldine",
		"Hunter",
	];

	let serverList;
	let allServers = [];
	try {
		serverList = await get("https://moomoo.io/serverData");
		serverList = serverList.substring(serverList.indexOf("{")).replace(";", "");
		serverList = JSON.parse(serverList);
		const servers = serverList.servers;
		let len = servers.length;
		let server;
		let games;
		let gameLen;
		while (len--) {
			server = servers[len];
			games = server.games;
			gameLen = games.length;
			while (gameLen--) {
				allServers.push({id: `${server.region}:${server.index}:${gameLen}`, ip: server.ip, gameIndex: gameLen});
			}
		}
	} catch (e) {
		console.error("Failed to load server data.");
		process.exit();
	}

	const bots = [];
	const tribes = {};
	const players = {};

	class Bot {
		constructor(n, ip, gameIndex, name, tribe, chatMsg, ai, probe, autoHeal, randSkins, hatID, autoAttack) {
			this.number = n;
			this.ip = ip;
			this.gameIndex = gameIndex;
			this.name = name || "unknown";
			this.tribe = tribe;
			this.chatMsg = chatMsg;
			this.origChatMsg = chatMsg;
			this.ai = ai;
			this.probe = probe;
			this.autoHeal = autoHeal;
			this.randSkins = randSkins;
			this.hatID = hatID;
			this.autoAttack = autoAttack;
			this.lastRandAngleUpdate = 0;
			this.pos = {
				x: 0,
				y: 0,
			};
			this.materials = {
				"wood": 0,
				"stone": 0,
				"food": 0,
				"points": 0,
			};
			this.chatInterval = undefined;
			this.reqint = undefined;
			this.updateInterval = undefined;
			this.key = null;
			this.id = null;
		}
		async connect() {
			const sk = this.socket = io.connect(`http://ip_${this.ip}.moomoo.io`, {
				reconnection: false,
				query: `gameIndex=${this.gameIndex}`,
				transportOptions: {
					polling: {
						extraHeaders: {
							"Origin": "http://moomoo.io",
							"Referer": "http://moomoo.io",
						},
					},
				},
				parser: parser,
			});
			if (this.probe) {
				sk.once("disconnect", () => {
					bots.splice(bots.indexOf(this), 1);
					this.socket.removeAllListeners();
					this.socket = null;
					if (bots.length === 0) {
						console.log("Probe finished.");
						process.exit();
					}
				});
				sk.once("connect", () => {
					bots.push(this);
					this.spawn();
				});
				// Spawn (id)
				sk.on("1", r => {
					this.id = r;
					setTimeout(this.socket.disconnect.bind(this.socket), 5000);
				});
				// Leaderboard
				sk.on("5", data => {
					if (probeName){
						if ((probeName instanceof RegExp && data.filter(d => typeof d === "string" && d.match(probeName)).length > 0) || (typeof probeName === "string" && data.indexOf(probeName) > -1)){
							console.log(`${getServerByIp(this.ip).id}`);
							sk.disconnect();
						}
					}
				});
				// ID (tribes[name, owner])
				sk.on("id", (data) => {
					data.teams.forEach(t => {
						if (probeTribe){
							if ((probeTribe instanceof RegExp && t.sid.match(probeTribe)) || (typeof probeTribe === "string" && t.sid === probeTribe)){
								console.log(`${getServerByIp(this.ip).id}`);
								sk.disconnect();
							}
						}
					});
				});
				return true;
			}else{
				sk.once("disconnect", () => {
					bots.splice(bots.indexOf(this), 1);
					console.log(`${this.number} disconnected`);
					clearInterval(this.chatInterval);
					clearInterval(this.reqint);
					clearInterval(this.updateInterval);
					setTimeout(this.connect.bind(this), 2000);
					this.socket.removeAllListeners();
					this.socket = null;
				});
				sk.once("connect", () => {
					console.log(`${this.number} connected`);
					bots.push(this);
					this.spawn();
				});
				// Spawn (id)
				sk.on("1", r => {
					console.log(`${this.number} spawned`);
					this.id = r;
					this.tribe && sk.emit("8", this.tribe);
					if (this.chatMsg) this.chatInterval = setInterval(this.chat.bind(this), 3000);
					this.tribe && (this.reqint = setInterval(this.join.bind(this), 5000));
					if (this.ai) this.updateInterval = setInterval(this.update.bind(this), 1000);
				});
				// Player Add ([l_id, id, name, x, y, angle, ?, ?, ?, ?], main)
				sk.on("2", (data, main) => {
					if (main) {
						this.longId = data[0];
						this.id = data[1];
						this.pos.x = data[3];
						this.pos.y = data[4];
						this.angle = data[5];
					}
					players[data[1]] = {
						longID: data[0],
						name: data[2],
						x: data[3],
						y: data[4],
						angle: data[5],
						lastUpdated: Date.now(),
					};
				});
				// Player Update ([id, x, y, angle, ?, ?, ?, tribe, ?, ?, ?, ?, ?])
				sk.on("3", (data) => {
					for (let i = 0, len = data.length / 13; i < len; i++) {
						if (this.id == data[0 + i * 13]) {
							this.pos.x = data[1 + i * 13];
							this.pos.y = data[2 + i * 13];
							this.angle = data[3 + i * 13];
						}
						if (!players[data[0 + i * 13]]) {
							players[data[0 + i * 13]] = {
								x: data[1 + i * 13],
								y: data[2 + i * 13],
								angle: data[3 + i * 13],
								lastUpdated: Date.now(),
							};
						}else{
							const p = players[data[0 + i * 13]];
							p.x = data[1 + i * 13];
							p.y = data[2 + i * 13],
							p.angle = data[3 + i * 13],
							p.lastUpdated = Date.now();
						}
					}
					this.update();
				});
				// Player Remove (l_id)
				sk.on("4", (longID) => {
					if (this === bots[0]) {
						for (const k in players) {
							if (this === bots[0]) {
								if (players[k].longID == longID) delete players[k];
							}
						}
					}
				});
				// Resource obtained
				sk.on("9", (type, amount) => {
					this.materials[type] = amount;
					this.tryHatOn(this.hatID);
				});
				// Damaged
				if (this.autoHeal) {
					sk.on("10", (id, health) => {
						if (id == this.id && health < 100) {
							setTimeout(this.heal.bind(this), 75 + (Math.random() / 10) | 0);
						}
					});
				}
				// Death
				sk.on("11", () => {
					console.log(`${this.number} died`);
					clearInterval(this.chatInterval);
					clearInterval(this.reqint);
					clearInterval(this.updateInterval);
					setTimeout(this.spawn.bind(this), 20);
				});
				// Tribe Delete (name)
				sk.on("ad", (name) => {
					if (this === bots[0]) {
						if (tribes[name]) delete tribes[name];
					}
				});
				// Tribe add request ({sid, owner})
				sk.on("ac", (data) => {
					tribes[name] = { owner: data.owner, players: [] };
				});
				// Tribe join request (id, name)
				sk.on("an", (id, name) => {
					sk.emit("11", id, 1);
					if (players[id]) {
						players[id].inMainTribe = true;
					}else{
						players[id] = { inMainTribe: true };
					}
				});
				// Chat (id, name)
				sk.on("ch", (id, msg) => {
					if (id != ownerID) return;
					if (!msg.startsWith("!")) return;
					const args = msg.slice(1).trim().split(/ +/g);
					const command = args.shift().toLowerCase();
					if (command === "fme") {
						goto.x = goto.y = null;
						stay = false;
						followID = ownerID;
						attackFollowedPlayer = false;
						followMouse = false;
					}else if (command === "id") {
						const a = [];
						for (const k in players) {
							if (players[k].name === args.join(" ")) a.push(k);
						}
						if (a.length > 0) {
							this.chatMsg = a.join(", ").slice(0, 30);
						}else{
							this.chatMsg = "Player not in memory.";
						}
						clearInterval(this.chatInterval);
						this.chatInterval = null;
						setTimeout(this.chat.bind(this), 1000);
					}else if (command === "fid") {
						goto.x = goto.y = null;
						stay = false;
						followID = parseInt(args[0]);
						attackFollowedPlayer = false;
						followMouse = false;
					}else if (command === "atkid") {
						goto.x = goto.y = null;
						stay = false;
						followID = parseInt(args[0]);
						attackFollowedPlayer = true;
						followMouse = false;
					}else if (command === "s") {
						goto.x = goto.y = null;
						stay = true;
						followID = null;
						attackFollowedPlayer = false;
						followMouse = false;
					}else if (command === "r") {
						goto.x = goto.y = null;
						stay = false;
						followID = null;
						attackFollowedPlayer = false;
						followMouse = false;
					}else if (command === "fm" && computer) {
						goto.x = goto.y = null;
						stay = false;
						followID = null;
						attackFollowedPlayer = false;
						followMouse = true;
					}else if (command === "hat" && args[0]) {
						const hatToEquip = args[0];
						let len = bots.length;
						let bot, triedHat;
						while (len--) {
							bot = bots[len];
							triedHat = bot.tryHatOn(hatToEquip);
							if (triedHat) {
								bot.chatMsg = "Switched hat.";
							}else if (triedHat === false) {
								bot.chatMsg = `Need ${data.hatPrices[getHatID(hatToEquip)] - bot.materials.points} more gold.`;
							}else{
								bot.chatMsg = "Invalid hat!";
							}
							clearInterval(bot.chatInterval);
							bot.chatInterval = null;
							setTimeout(bot.chat.bind(bot), 1000);
						}
					}else if (command === "atk") {
						this.autoAttack = !this.autoAttack;
						this.socket && this.socket.emit("7", this.autoAttack);
					}else if (command === "sp") {
						this.socket.emit("5", 5, null);
						this.socket.emit("4", 1, null);
						this.socket.emit("4", 0, null);
						this.socket.emit("5", 1, null);
					}else if (command === "w") {
						this.socket.emit("5", 2, null);
						this.socket.emit("4", 1, null);
						this.socket.emit("4", 0, null);
						this.socket.emit("5", 1, null);
					}
				});
				// ID (tribes[name, owner])
				sk.on("id", (data) => {
					if (this === bots[0]) {
						data.teams.forEach(t => {
							tribes[t.sid] = { owner: t.owner, players: [] };
						});
					}
				});
				// Tribe Update [id_1, name_1, ...]
				sk.on("sa", (data) => {
					if (tribes[this.tribe]) {
						tribes[this.tribe].players = [];
						for (let i = 0, len = data.length; i < len; i += 2) {
							tribes[this.tribe].players.push({ id: data[i], name: data[i + 1] });
						}
					}
				});
				// Tribe Start (name, owner)
				sk.on("st", (name, owner) => {
					if (this === bots[0]) {
						tribes[name] = { owner: owner, players: [] };
					}
				});
				// Minimap Ping (x, y)
				sk.on("p", (x, y) => {
					goto.x = x;
					goto.y = y;
				});
				return true;
			}
		}
		disconnect() {
			this.socket && this.socket.disconnect();
		}
		spawn() {
			this.socket && this.socket.emit("1", {
				name: this.name,
				moofoll: true,
				skin: this.randSkins ? Math.round(Math.random() * 5) : 0,
			});
			this.socket && this.socket.emit("7", this.autoAttack);
			this.tryHatOn(this.hatID);
		}
		join() {
			this.socket && this.tribe && this.socket.emit("10", this.tribe);
		}
		heal() {
			if (!this.socket) return;
			this.socket.emit("5", 0, null);
			this.socket.emit("4", 1, null);
			this.socket.emit("5", 0, true);
		}
		tryHatOn(id) {
			if (!this.socket) return;
			id = getHatID(id);
			if (isNaN(id)) return null;
			if (!isNaN(data.hatPrices[id])) {
				this.hatID = id;
				if (this.materials.points >= data.hatPrices[id]) {
					this.socket.emit("13", 1, id);
					this.socket.emit("13", 0, id);
					return true;
				}else{
					this.socket.emit("13", 0, id);
					return false;
				}
			}
		}
		chat() {
			if (!this.socket) return;
			this.socket.emit("ch", this.chatMsg);
			this.chatMsg = this.origChatMsg;
			if (this.origChatMsg && !this.chatInterval) {
				this.chatInterval = setInterval(this.chat.bind(this), 3000);
			}
		}
		update() {
			if (!this.socket) return;
			if (!this.ai) return;
			if (stay) {
				this.socket.emit(3, null);
			}else if (followMouse) {
				const p = players[ownerID];
				if (p && p.x) {
					const targetX = p.x + (mousePos.x - screen.width / 2);
					const targetY = p.y + (mousePos.y - screen.height / 2);
					if (Math.pow(this.pos.x - targetX, 2) + Math.pow(this.pos.y - targetY, 2) < 20000) {
						this.socket.emit(2, p.angle);
						this.socket.emit(3, null);
					}else{
						this.socket.emit(2, Math.atan2(targetY - this.pos.y, targetX - this.pos.x));
						this.socket.emit(3, Math.atan2(targetY - this.pos.y, targetX - this.pos.x));
					}
				}
			}else if (followID && players[followID]) {
				const p = players[followID];
				if (p && p.x) {
					const now = Date.now();
					if (now - p.lastUpdated > 30000 && now - this.lastRandAngleUpdate > 20000) {
						this.lastRandAngleUpdate = now;
						const randAngle = Math.random() * Math.PI * 2;
						this.socket.emit(2, randAngle);
						this.socket.emit(3, randAngle);
					}else if (!attackFollowedPlayer) {
						if (Math.pow(this.pos.x - p.x, 2) + Math.pow(this.pos.y - p.y, 2) < 20000) {
							this.socket.emit(2, p.angle);
							this.socket.emit(3, null);
						}else{
							this.socket.emit(2, Math.atan2(p.y - this.pos.y, p.x - this.pos.x));
							this.socket.emit(3, Math.atan2(p.y - this.pos.y, p.x - this.pos.x));
						}
					}else{
						this.socket.emit(2, Math.atan2(p.y - this.pos.y, p.x - this.pos.x));
						this.socket.emit(3, Math.atan2(p.y - this.pos.y, p.x - this.pos.x));
					}
				}
			}else if (goto.x && goto.y) {
				if (Math.pow(this.pos.x - goto.x, 2) + Math.pow(this.pos.y - goto.y, 2) < 40000) {
					this.socket.emit(2, 0);
					this.socket.emit(3, null);
				}else{
					this.socket.emit(2, Math.atan2(goto.y - this.pos.y, goto.x - this.pos.x));
					this.socket.emit(3, Math.atan2(goto.y - this.pos.y, goto.x - this.pos.x));
				}
				return;
			}else{
				this.socket.emit(3, null);
			}
		}
	}

	function escapeRegExp(s) {
		if (s.startsWith("/") && s.endsWith("/")){
			s = s.split("");
			s = s.slice(1, s.length - 1).join("");
		}
	  return s.replace(/[-/\\^$*+?.()|[\]{}]/g, "\\$&");
	}

	const numBots = Math.min((args.num && parseInt(args.num.value)) || 0, 0b01010000);
	const link = (args.link && getIP(args.link.value)) || null;
	let name = (args.randNames && args.randNames.value.toLowerCase() != "false" && args.randNames.value != 0) ? true : ((args.name && args.name.value) || "unknown");
	let tribe = (args.tribe && args.tribe.value) || null;
	let chat = (args.chat && args.chat.value) || null;
	const ai = args.ai && args.ai.value.toLowerCase() != "false" && args.ai.value.toLowerCase() != "0";
	const probeRegex = args.probeRegex && args.probeRegex.value.toLowerCase() != "false" && args.probeRegex.value.toLowerCase() != "0";
	const probeTribe = args.probeTribe && (probeRegex ? new RegExp(escapeRegExp(args.probeTribe.value)) : args.probeTribe.value);
	const probeName = args.probeName && (probeRegex ? new RegExp(escapeRegExp(args.probeName.value)) : args.probeName.value);
	const probe = probeTribe || probeName;
	const autoHeal = !args.autoHeal || (args.autoHeal.value.toLowerCase() != "false" && args.autoHeal.value.toLowerCase() != "0");
	const randSkins = args.randSkins && args.randSkins.value.toLowerCase() != "false" && args.randSkins.value.toLowerCase() != "0";
	const hatID = (args.hat && args.hat.value) || null;
	const autoAttack = !args.autoAttack || (args.autoAttack.value.toLowerCase() != "false" && args.autoAttack.value.toLowerCase() != "0");
	typeof name === "string" && (name = name.slice(0, 16));
	tribe && (tribe = tribe.slice(0, 6));
	chat && (chat = chat.slice(0, 30));

	if (probe) {
		if (probeRegex){
			console.log(`Initiating probe for${(args.probeTribe && args.probeTribe.value) ? ` tribe ${args.probeTribe.value}` : ""}${(args.probeName && args.probeName.value) ? ` player ${args.probeName.value}` : ""} using regex.`);
		}else{
			console.log(`Initiating probe for${probeTribe ? ` tribe ${probeTribe}` : ""}${probeName ? ` player ${probeName}` : ""}.`);
		}
		(function connectBots(i) {
			if (i <= 0) return;
			const promises = [];
			for (let j = i; (j > i - 8) && (j > 0); j--) {
				promises.push(new Bot(j, allServers[j - 1].ip, allServers[j - 1].gameIndex, "PROBE", tribe, chat, ai, probe, autoHeal, randSkins, hatID, autoAttack).connect());
			}
			Promise.all(promises).then(() => {
				connectBots(i - 8);
			}).catch(console.error);
		})(allServers.length);
	} else {
		if (numBots > 20) {
			process.stdout.write("Warning: 20 or more bots may be capped by a server-side IP limit.\n");
		}

		const server = getServerByID(link);
		if (!server) {
			console.error("Invalid link!");
			process.exit();
		}
		(function connectBots(i) {
			if (i <= 0) return;
			const promises = [];
			for (let j = i; (j > i - 8) && (j > 0); j--) {
				promises.push(new Bot(j, server.ip, server.gameIndex, name === true ? names[(Math.random() * names.length) | 0] : name, tribe, chat, ai, probe, autoHeal, randSkins, hatID, autoAttack).connect());
			}
			Promise.all(promises).then(() => {
				connectBots(i - 8);
			}).catch(console.error);
		})(numBots);
	}

	process.stdin.setEncoding("utf8");
	process.stdin.on("data", data => {
		data.split(/[\r\n]+/).forEach(l => {
			processInput(l.trim());
		});
	});
	process.stdin.resume();
})();
